Sonarr: avoid downgrade on package updates & version bump > Closed to #5515

mreid-tt opened the pull request:
## Description

This builds on the work from #5485 to address a number of issues in Sonarr as discussed in https://github.com/SynoCommunity/spksrc/pull/5471#issuecomment-1317381973 and summarised below:

1. Avoid Sonarr downgrade on package update
2. Upgrade to Sonarr 3.0.9.1549
3. Implement fix for alternate TMP directory
4. Implement 'config.xml' but preserve other keys in distribution

Fixes #5450

## Checklist

- [x] Build rule `all-supported` completed successfully
- [x] New installation of package completed successfully
- [x] Package upgrade completed successfully (Manually install the package again)
- [x] Package [functionality was tested](https://github.com/SynoCommunity/spksrc/wiki/Update-Policy#tests-checks)
- [x] Any needed [documentation](https://github.com/SynoCommunity/spksrc/wiki/Create-documentation) is updated/created


### Type of change

<!--Please use any relavent tags.-->
- [x] Bug fix
- [x] Package update

mreid-tt commented:
@hgy59, I've gone ahead and created this updated PR to replace #5471. I did notice however when I included a `config.xml` similar to the one you have in [spksrc/spk/lidarr/src](https://github.com/SynoCommunity/spksrc/blob/master/spk/lidarr/src/config.xml), I get an error in the logs during the upgrade:

```
mv: cannot stat '/volume1/@appstore/nzbdrone/app/config.xml': No such file or directory
```

What I found however is that even without this file, the automatic update on launch of Sonarr was in fact executed. I realise that this was because of the presence of the file: `update_required`.

On further investigation and comparing with #5485, I realise that the automatic update does not in fact fire when this line is in the `config.xml`:

https://github.com/SynoCommunity/spksrc/blob/44a3d5d31002250a5fa66281e392f671275e2e95/spk/lidarr/src/config.xml#L5

This was confirmed by taking the following steps in Lidarr:

1. Manually perform a clean install of Lidarr version 20210322-7 from [SynoCommunity](https://synocommunity.com/package/lidarr)
2. Upgrade to current version 20221114-10 from Package Center
3. Observe that the automatic update does not fire and that the file `.../lidarr/.config/Lidarr/update_required` is still present
4. Within the Lidarr app, uncheck 'Automatic' from [Settings > General > Updates]
5. Restart Lidarr from Package Center
6. Confirm that the automatic update fires and the file (in step 3 above) is not present

As such, I have removed this configuration from this PR as it is not required. Further, I would recommend that a future PR, remove the 'UpdateAutomatically' setting from the `config.xml` for Lidarr.

hgy59 commented:
> As such, I have removed this configuration from this PR as it is not required. Further, I would recommend that a future PR, remove the 'UpdateAutomatically' setting from the `config.xml` for Lidarr.

My finding was, that without `UpdateAutomatically=true` the `update_required` file was not processed in Lidarr, but I have to double check this.

mreid-tt commented:
Okay @hgy59, fair enough. Let me know if you want me to create a Lidarr PR for testing that. As for this PR, I hope you can review and confirm if you are comfortable with the changes proposed. Once done, #5471 can be closed as it would no longer be required. Looking forwarding to fixing the longstanding issue in #5450.

bakerboy448 commented:
As previously discussed on the Prowlarr GH - all the arrs (until recently - and thus the future new masters) do not respect `update_required`

bakerboy448 commented on spk/sonarr/src/service-setup.sh:
```diff
@@ -43,62 +36,69 @@ SVC_WAIT_TIMEOUT=90
+    echo "Set update required"
+    # Make Sonarr do an update check on start to avoid possible Sonarr
+    # downgrade when synocommunity package is updated
+    touch "${SONARR_CONFIG_DIR}/update_required"
```
this requires the config file to be set to True for update automatically to function.

```csharp
                // Don't do a prestartup update check unless BuiltIn update is enabled
                if (!_configFileProvider.UpdateAutomatically ||
                    _configFileProvider.UpdateMechanism != UpdateMechanism.BuiltIn ||
                    _deploymentInfoProvider.IsExternalUpdateMechanism)
                {
                    _logger.Debug("Built-in updater disabled, skipping post-install update check");
                    return;
                }
```

https://github.com/Sonarr/Sonarr/blob/main/src/NzbDrone.Core/Update/InstallUpdateService.cs#L304-L311

the config file default is false

```csharp
        public bool UpdateAutomatically => GetValueBoolean("UpdateAutomatically", false, false);
```
https://github.com/Sonarr/Sonarr/blob/e4c0e80e3e12688a2712078b65d9693c9e3d5c8f/src/NzbDrone.Core/Configuration/ConfigFileProvider.cs#L223

bakerboy448 reviewed (changes requested):
the config file to set attempts to launch browser to false is also required.

mreid-tt commented:
thanks @bakerboy448 for the detailed feedback. Based on the above it would seem that the `config.xml` is required; however, based on the code you sent the only parameter needed is `UpdateMechanism`. Having this set to 'BuiltIn' would be a pre-requisite to trigger the function for the `update_required` file. Checking in another PR you further mentioned in https://github.com/SynoCommunity/spksrc/pull/5456#issuecomment-1295887770:

> Additionally, Starr apps do not automatically update on install unless both update_required exists is in the appdata directory AND the app is configured to automagically update (UpdateAutomatically is false by default)

This has not been my experience in practice since it does not seem to automatically update for me when `update_required` exists and `UpdateAutomatically` is set to true. I carried out the following experiment:

1. Clean install Sonarr version 20210329-19 from SynoCommunity
2. Stop the Sonarr app
3. Manually upgrade to version 20221204-21 from this PR
4. Move the `config.xml` into the Sonarr AppData directory
5. Start the Sonarr app
6. Observe that the internal version is still at 3.0.5.1144 (even after a second restart)
7. Stop the Sonarr app
8. Remove the line for 'UpdateAutomatically' in the `config.xml` file
9. Start the Sonarr app
10. Confirm that the automatic updater fired and the internal version upgraded to 3.0.9.1549

For the above experiment the following data was used for the `config.xml` in step 4:

```
<Config>
  <UpdateMechanism>BuiltIn</UpdateMechanism>
  <UpdateAutomatically>True</UpdateAutomatically>
  <LaunchBrowser>False</LaunchBrowser>
  <Branch>main</Branch>
</Config>
```

Based on the above the automatic update does not seem to fire when the 'UpdateAutomatically' parameter in the `config.xml` is set to true. As such, I am inclined to include only the following data in the `config.xml`:

```
<Config>
  <UpdateMechanism>BuiltIn</UpdateMechanism>
  <LaunchBrowser>False</LaunchBrowser>
  <Branch>main</Branch>
</Config>
```

What are your thoughts about this?

On a different note, I have two potential challenges with including the `config.xml`. Based on my earlier tests, the installation does not seem to be able to locate it during the post-install and I get the following error:

```
mv: cannot stat '/volume1/@appstore/nzbdrone/app/config.xml': No such file or directory
```

I am not sure how to fix this. I had the `config.xml` file in the same location as in Lidarr and Radarr and the line to move it in the code seems to be identical but for some reason it's not finding the file. Any thoughts on this?

Finally, if we are replacing the `config.xml` file on install or upgrade, this may pose a problem for existing installations who may have existing configurations which would be erased. One obvious issue is the re-generation of a new API key. Would there be a way to just add or replace the specific parameters we want changed if a `config.xml` already exists?

bakerboy448 commented:
> This has not been my experience in practice since it does not seem to automatically update for me when update_required exists and UpdateAutomatically is set to true. I carried out the following experiment:

that shouldn't be the case and would be a bug to report.

> but for some reason it's not finding the file. Any thoughts on this?

that'd be a question for the syno folks here for why that's the case.

> if we are replacing the config.xml file on install or upgrade

why would you be overwriting the file if it exists? that seems like poor practice.

> or replace the specific parameters we want changed if a config.xml already exists?

just use sed (assuming it exists on synology) to update values as needed or rather if that's the case, why mess with the config at all?

mreid-tt commented:
Hi @bakerboy448, appreciate the responses. I think I found some further information...

> > This has not been my experience in practice since it does not seem to automatically update for me when update_required exists and UpdateAutomatically is set to true. I carried out the following experiment:
> 
> that shouldn't be the case and would be a bug to report.

I've looked into this and it seems that based on the history of 'InstallUpdateService.cs', the logic for the updater was fixed in https://github.com/Sonarr/Sonarr/commit/eea6be459d5dbfafb9a5285046282c25c4697242 which was pushed on Mar 14, 2021 however the release for Sonarr 3.0.5.1144 came out prior on Mar 7, 2021 so there was in fact a bug at that time.

As such for simplicity, I will include the `UpdateAutomatically` as you have instructed.

> > but for some reason it's not finding the file. Any thoughts on this?
> 
> that'd be a question for the syno folks here for why that's the case.

@hgy59, any thoughts as to why I'm getting this issue with Sonarr?

> > if we are replacing the config.xml file on install or upgrade
> 
> why would you be overwriting the file if it exists? that seems like poor practice.
> 
> > or replace the specific parameters we want changed if a config.xml already exists?
> 
> just use sed (assuming it exists on synology) to update values as needed or rather if that's the case, why mess with the config at all?

Thanks for this suggestion. It does make more sense to edit the file if it is present. I'll take a look at 'sed' and re-write the PR to include this logic.

mreid-tt commented:
Okay, I believe all remaining issues have been resolved and tested. In summary the following were included in the final squashed commits:

1. Added code to intelligently add or modify key values in the `config.xml` file if present
2. Updated the `Makefile` to copy the 'config.xml' file to the correct path (similar to Radarr/Lidarr)

With these changes included, the installation works well for both a fresh install and an upgrade install without any of the previously reported issues. The package should be ready for merging; do let me know if there are any further concerns @bakerboy448.

bakerboy448 commented on spk/sonarr/Makefile:
```diff
@@ -16,7 +16,7 @@ DESCRIPTION_FRE = Sonarr est un PVR pour les utilisateurs de groupes de discussi
 DISPLAY_NAME = Sonarr
 STARTABLE = yes
-CHANGELOG = "Enlarge the service start/stop timeout to 90 seconds."
+CHANGELOG = "1. Enlarge the service start/stop timeout to 90 seconds<br/>2. Avoid Sonarr downgrade on package update<br/>3. Upgrade to Sonarr 3.0.9.1549"
```
The sonarr service should also be changed to use the alt temp directory like radarr/lidarr etc given that DSM 7 restrictions on /tmp and general future proofing

mreid-tt replied:
Thanks for pointing this out. I'll update the PR to include the alternate temp directory.

mreid-tt commented on spk/sonarr/Makefile:
```diff
@@ -35,6 +35,11 @@ POST_STRIP_TARGET = sonarr_extra_install
 include ../../mk/spksrc.spk.mk
 
+PACKAGE_VERSION = $(SPK_NAME_ARCH)-$(SPK_TCVERS)_$(SPK_VERS)-$(SPK_REV)
+PACKAGE_AUTHOR = [Team Sonarr](https://sonarr.tv)
```
I was just going based on these examples:
https://github.com/SynoCommunity/spksrc/blob/44a3d5d31002250a5fa66281e392f671275e2e95/spk/lidarr/Makefile#L39-L40
https://github.com/SynoCommunity/spksrc/blob/44a3d5d31002250a5fa66281e392f671275e2e95/spk/radarr/Makefile#L43-L44
So I'm guessing they were going for author and nor maintainer?

mreid-tt replied:
So what would you propose? Something like this?
```
PACKAGE_AUTHOR = [SynoCommunity](https://synocommunity.com)
```

bakerboy448 replied:
That's up to SynoComm

mreid-tt replied:
@hgy59, can you weigh in on this please? Or tag whomever else should comment on this.

mreid-tt replied:
I've reviewed how this presents in the interface and I believe my proposal should be fine. Updated the PR.

bakerboy448 replied:
What's that screenshot look like? Since again the Sonarr team neither maintain, support, nor develop the package.

mreid-tt replied:
While in the other ***arr** apps, the information in the `package_info` file seems to be added to the system status screen, the code to pull this information does not seem to be present in Sonarr. As such, the screenshot looks no different with or without this file... Here's what it looks like from this PR:

<img width="484" alt="Screenshot 2022-12-07 at 5 29 31 PM" src="https://user-images.githubusercontent.com/943378/206300522-18edac31-d1f2-4c0d-a28d-294f5eef4b32.png">

And here's what the other packages look like that have this code:

<img width="492" alt="Screenshot 2022-12-07 at 5 30 52 PM" src="https://user-images.githubusercontent.com/943378/206300756-29636f37-0a57-4592-92d5-50294a75bd20.png">
<img width="482" alt="Screenshot 2022-12-07 at 5 31 00 PM" src="https://user-images.githubusercontent.com/943378/206300766-2b5c59eb-f33d-4aa4-8606-97c35f64ff66.png">

bakerboy448 replied:
That info comes from the package_info file.

What does the Syno Package itself look like in store / when adding or whatever?

mreid-tt replied:
The only other place I see that code is perhaps in the installation screen like this:

<img width="689" alt="Screenshot 2022-12-07 at 7 59 05 PM" src="https://user-images.githubusercontent.com/943378/206323273-17a485a6-ae2f-4d5b-81dd-cfc1995cf654.png">

This would be different than with another package like Lidarr:

<img width="686" alt="Screenshot 2022-12-07 at 8 00 03 PM" src="https://user-images.githubusercontent.com/943378/206323303-be2ad70b-3098-42e2-a1e2-b717081fa8ff.png">

mreid-tt replied:
But to be fair, the screen looks the same as the Sonarr currently in the SynoCommunity repository:

<img width="689" alt="Screenshot 2022-12-07 at 8 03 34 PM" src="https://user-images.githubusercontent.com/943378/206323740-1155be89-5e58-49f3-8bf4-02a1b3389452.png">

So I'm not really sure where the `package_info` would make a difference.

mreid-tt commented on spk/sonarr/src/service-setup.sh:
```diff
@@ -43,62 +36,103 @@ SVC_WAIT_TIMEOUT=90
+            values+=("$value")
+        done < "${SYNOPKG_PKGDEST}/app/config.xml"
+        # Iterate through the keys and values
+        for ((i = 0; i < ${#keys[@]}; i++)); do
```
I may have missed something as i'm not clear on the concern here. I don't believe I have any lines printing the keys or vales to logs. Can you elaborate which log file you are concerned with?

The logic of the operation is to first read in the keys and values from the XML file in the package ([spk/sonarr/src/config.xml](https://github.com/SynoCommunity/spksrc/blob/380ba22804361fd635dd68ab7d818afe1c3bff03/spk/sonarr/src/config.xml)) and then search for these keys in the `config.xml` from the installed distribution. The keys and values are stored in a temporary array and the only reason that 'ApiKey' would be read is if it is present in the XML file in the package.

mreid-tt commented:
> > This was confirmed by taking the following steps in Lidarr:
> > 
> > 1. Manually perform a clean install of Lidarr version 20210322-7 from [SynoCommunity](https://synocommunity.com/package/lidarr)
> > 2. Upgrade to current version 20221114-10 from Package Center
> > 3. Observe that the automatic update does not fire and that the file `.../lidarr/.config/Lidarr/update_required` is still present
> > 4. Within the Lidarr app, uncheck 'Automatic' from [Settings > General > Updates]
> > 5. Restart Lidarr from Package Center
> > 6. Confirm that the automatic update fires and the file (in step 3 above) is not present
> > 
> > As such, I have removed this configuration from this PR as it is not required. Further, I would recommend that a future PR, remove the 'UpdateAutomatically' setting from the `config.xml` for Lidarr.
> 
> My finding was, that without `UpdateAutomatically=true` the `update_required` file was not processed in Lidarr, but I have to double check this.

@hgy59, just to close the loop on this, I reconfirmed that my testing methodology may have been a contributor to a different finding than yours. One of the causes it would not work is because of the following:

1. For the clean install of Lidarr version 20210322-7, the internal version was 0.8.0.2042
2. Based on the [release history](https://github.com/Lidarr/Lidarr/releases), this version was published on Jan 28, 2021
3. The code present in [`.../Update/InstallUpdateService.cs`](https://github.com/Lidarr/Lidarr/blob/develop/src/NzbDrone.Core/Update/InstallUpdateService.cs), had an updater fix on Mar 21, 2021
4. This commit was titled [Fixed post-install update check not running](https://github.com/Lidarr/Lidarr/commit/06d2a297eba43796b096a5f92be9c27a8f6ea9bf) which corrected a logic bug
5. This logic bug would still be present on upgrade to version 20221114-10 from Package Center since the current distribution is retained

This was the likely cause of the issue I observed. It is noted however that if I do a clean install of version 20221114-10 from Package Center (with internal version 1.0.2.2592 that does not have the updater bug), the file `.../lidarr/.config/Lidarr/update_required` is still present and not removed correctly. This may be related to the comment from @bakerboy448 where he mentioned:

> As previously discussed on the Prowlarr GH - all the arrs (until recently - and thus the future new masters will be the ones that will have the fix) do not respect `update_required` due to an inadvertent regression.

This I believe is what I observe happening now but this is not the case with Sonarr at the moment. A separate bug report will likely be needed for the Lidarr regression.

bakerboy448 commented:
No bug report for lidarr is needed as it's already fixed in nightly along with the other \*arrs.

hgy59 commented:
> This was the likely cause of the issue I observed. It is noted however that if I do a clean install of version 20221114-10 from Package Center (with internal version 1.0.2.2592 that does not have the updater bug), the file `.../lidarr/.config/Lidarr/update_required` is still present and not removed correctly. This may be related to the comment from @bakerboy448 where he mentioned:

If one has installed version 20221114-10 it needs definitely a manual update. An update with 20221204-21 (or any newer version) cannot automatically update from an earlier version that has not the sonarr autoupdate fix. 
It will only be fixed in the future, when you install the new package and a future version is available or already installed.

So there are only two scenarios now, to test whether the auto update works:

A)
1. install 20221114-10 
2. manually update inside sonnar 
3. "update" with package 20221114-10
4. The `update_required` file must be deleted

B)
1. Uninstall sonarr
2. install package 20221114-10
3. The `update_required` file must be deleted

Scenario in the future, when a newer version than in package 20221114-10 is available
1. Uninstall sonarr
2. install package 20221114-10
3. The `update_required` file is deleted
4. sonarr is automatically updated to the available version

This is by design of our packages with enabled internal updaters. It is important to avoid downgrade on updates, as it could break the packages. 
It is by logic that the automatic update will only work, when it works with the installed version.

BTW 
my findings in Lidarr were, that in "B) 3." the file was deleted only with `<UpdateAutomatically>True</UpdateAutomatically>` in config.xml.

mreid-tt commented:
@hgy59, thanks for the feedback and analysis. I believe we are on the same page that the 'config.xml' should include the automatic update as shown:
https://github.com/SynoCommunity/spksrc/blob/0a8440479e3cb76b2d119751b6b944cbcbc5ff65/spk/sonarr/src/config.xml#L5
Can this PR be given a final review to determine if there are any other points to resolve before merging and publishing?

hgy59 commented:
> Can this PR be given a final review to determine if there are any other points to resolve before merging and publishing?

Sorry for beeing late with a deeper review...

- I propose to use the same backup/restore implementation as in radarr, leadarr et al. without evaluation of the installed version. i.e. just backup and restore the folder "share"
- I propose not to touch the config.xml file on package updates. Why not use the target config file folder for config.xml in the package? (i.e. var/.config/Sonarr/config.xml instead of app/config.xml) and not merging the config.xml file in postinst (the whole var folder is backuped and restored by the generic installer code und updates). AND we should not change the config of an existing installation. If a user has disabled UpdateAutomatically, an update must not change this!
- And I still vote for removing the v2 to v3 migration code in the installer. If it must be expected, that anyone could try to update from Sonarr 2.x, we could check this in `validate_preinst` and abort the installation with an error message.


BTW: Please create a dedicated branch on your fork for each PR. Without such branches, you will get serious troubles with your master branch.

mreid-tt commented:
Hey thanks for the evaluation, some responses for you:

> * I propose to use the same backup/restore implementation as in radarr, leadarr et al. without evaluation of the installed version. i.e. just backup and restore the folder "share"

We do actually use the same backup and restore if we are running Sonarr v3. The only evaluation that is done is if the user is coming from v2 and if they are, the package binaries would be used.

> * I propose not to touch the config.xml file on package updates. Why not use the target config file folder for config.xml in the package? (i.e. var/.config/Sonarr/config.xml instead of app/config.xml) and not merging the config.xml file in postinst (the whole var folder is backuped and restored by the generic installer code und updates). AND we should not change the config of an existing installation. If a user has disabled UpdateAutomatically, an update must not change this!

When I looked at the code that is in Lidarr and Radarr, it actually replaces the user's 'config.xml' in their distribution which is not good (since for example it wipes their ApiKey currently). So as a better option, I slimmed the 'config.xml' to just the attributes we would need to trigger an update on the next run (`UpdateAutomatically`, `UpdateMechanism` and the update type with `Branch`). The only other key that is set would be `LaunchBrowser` since deployments prior to this one did not disable it.

If we don't touch the 'config.xml' in the existing distribution, are we okay with it not triggering an update on deployment as well as leaving the `LaunchBrowser` enabled? If so then I can happily remove all the complexity I added to achieve the above.

> * And I still vote for removing the v2 to v3 migration code in the installer. If it must be expected, that anyone could try to update from Sonarr 2.x, we could check this in `validate_preinst` and abort the installation with an error message.

If we are going in this direction, I need some more information to be able to author that.

> BTW: Please create a dedicated branch on your fork for each PR. Without such branches, you will get serious troubles with your master branch.

My apologies. I don't know if I can make a branch and move all this to at this point but I'll look into it.

hgy59 commented:
> When I looked at the code that is in Lidarr and Radarr, it actually replaces the user's 'config.xml' in their distribution

Yes it is installed, but finally replaced by the previously installed version 

This is different for DSM6 and DSM7:

DSM6:
see `preupgrade`  and `postupgrade` funtions in `spksrc.service.installer.dsm6`

in preupgrade the following line makes a backup of the var folder to $(TMP_DIR)
```
$CP "${SYNOPKG_PKGVAR}"/. "$TMP_DIR" 2>&1 | install_log
```
and in postupgrde the following line restores the whole var folder:
```
$CP "${TMP_DIR}"/. "${SYNOPKG_PKGVAR}" 2>&1 | install_log
```

DSM7:
the var folder is not overwritten by the var folder in the spk package.
The generic installer (see `spksrc.service.installer.dsm7`) synchronizes new files from the package to the var folder. Alredy existing files are not overwridden and files from the package with conflicting names are installed into the var folder under a different name (*.new), see `syno_sync_var_folder` function.

Packages for DSM 7 have special handling when the former package installation was done on DSM 6. The var folder will be moved to the new location. This is also implemented in the generic installer (see preupgrade and postupgrade in  `spksrc.service.installer.dsm7`)

And on DSM 7 the var folder is not removed on uninstall (except you select so in the DSM 7 uninstaller of our packages).

hgy59 commented:
> > * And I still vote for removing the v2 to v3 migration code in the installer. If it must be expected, that anyone could try to update from Sonarr 2.x, we could check this in `validate_preinst` and abort the installation with an error message.
> 
> If we are going in this direction, I need some more information to be able to author that.


The validation would be like this:

define a `validate_preinst` function in `service-setup.sh`
in this function evaluate the installed version 
for version 2.x output and error text with echo and finally call `exit 1`
otherwise do nothing.


You find a skeleton for such implementation in `spk/demoservice/src/service-setup.sh`
you can play around with demoservice (it builds fast).

mreid-tt commented:
Thanks for the feedback but I have a challenge to this as it has not been my experience in DSM7:

> > When I looked at the code that is in Lidarr and Radarr, it actually replaces the user's 'config.xml' in their distribution
> 
> Yes it is installed, but finally replaced by the previously installed version

So I tried to test the above with Lidarr and Radarr again using the versions prior to the current one in SynoCommunity and then upgrading to the current release.

### Lidarr

For this I used Version 20210322-7 from https://synocommunity.com/package/lidarr to perform a clean install for my platform. Once installed and run a `config.xml` file was created at this path: '/volume1/@appdata/lidarr/.config/Lidarr/'. The file contained the following:

```
<Config>
  <Branch>develop</Branch>
  <UpdateMechanism>BuiltIn</UpdateMechanism>
  <LaunchBrowser>False</LaunchBrowser>
  <Port>8686</Port>
  <SslPort>8687</SslPort>
  <LogLevel>info</LogLevel>
  <BindAddress>*</BindAddress>
  <EnableSsl>False</EnableSsl>
  <SslCertPath></SslCertPath>
  <UrlBase></UrlBase>
  <ApiKey>26c60e6f0c2d4f98badc63b2807923e8</ApiKey>
  <AuthenticationMethod>None</AuthenticationMethod>
</Config>
```

I then proceeded to upgrade via Package Center to the current Version 20221114-10. Once this process was completed the `config.xml` in the path '/volume1/@appdata/lidarr/.config/Lidarr/' looks like this:

```
<Config>
  <Branch>master</Branch>
  <UpdateMechanism>BuiltIn</UpdateMechanism>
  <UpdateAutomatically>True</UpdateAutomatically>
  <LaunchBrowser>False</LaunchBrowser>
  <Port>8686</Port>
  <LogLevel>info</LogLevel>
  <BindAddress>*</BindAddress>
  <EnableSsl>False</EnableSsl>
  <SslCertPath></SslCertPath>
  <UrlBase></UrlBase>
  <ApiKey>d697cd09d7a34c8db5627db56ff2b2d1</ApiKey>
  <AuthenticationMethod>None</AuthenticationMethod>
</Config>
```

As you will note in the above, keys such as `Branch` and `ApiKey` are decidedly different and a new key `UpdateAutomatically` is present. When I first observed this behaviour I attributed this to the following code in the `service_postinst` funtion:

https://github.com/SynoCommunity/spksrc/blob/9addc7905cda7c2c3686ac5576168d73c6244c0c/spk/lidarr/src/service-setup.sh#L23

### Radarr

For this I used Version 20210311-15 from https://synocommunity.com/package/radarr to perform a clean install for my platform. Once installed and run a `config.xml` file was created at this path: '/volume1/@appstore/radarr/var/.config/Radarr/'. The file contained the following:

```
<Config>
  <Branch>master</Branch>
  <UpdateMechanism>BuiltIn</UpdateMechanism>
  <LaunchBrowser>False</LaunchBrowser>
  <Port>8310</Port>
  <SslPort>8311</SslPort>
  <LogLevel>info</LogLevel>
  <BindAddress>*</BindAddress>
  <EnableSsl>False</EnableSsl>
  <SslCertPath></SslCertPath>
  <UrlBase></UrlBase>
  <ApiKey>101c31bc06d447a391c5ab1cd32f7ee6</ApiKey>
  <AuthenticationMethod>None</AuthenticationMethod>
</Config>
```

I then proceeded to upgrade via Package Center to the current Version 20220515-18. Once this process was completed the `config.xml` in the path '/volume1/@appdata/radarr/.config/Radarr/' looks like this:

```
<Config>
  <Branch>master</Branch>
  <UpdateMechanism>BuiltIn</UpdateMechanism>
  <LaunchBrowser>False</LaunchBrowser>
  <Port>8310</Port>
  <SslPort>8311</SslPort>
  <LogLevel>info</LogLevel>
  <UrlBase></UrlBase>
  <AuthenticationMethod>None</AuthenticationMethod>
  <ApiKey>7690839bf64b4b18b9bc7ee1c929c23e</ApiKey>
  <InstanceName>Radarr</InstanceName>
</Config>
```

As you will note in the above, the key `ApiKey` is decidedly different and a new key `InstanceName` is present. When I first observed this behaviour I attributed this to the following code in the `service_postinst` funtion:

https://github.com/SynoCommunity/spksrc/blob/9addc7905cda7c2c3686ac5576168d73c6244c0c/spk/radarr/src/service-setup.sh#L25

### Conclusion

> DSM7:
> the var folder is not overwritten by the var folder in the spk package.

Based on the above observation, I don't think this is working as proposed. Should we be looking to resolve this bug before moving forward? Eager to hear your thoughts on this.

hgy59 commented:
> Based on the above observation, I don't think this is working as proposed. Should we be looking to resolve this bug before moving forward? Eager to hear your thoughts on this.

@mreid-tt thanks for drill down this issue!

Simple answer: it was not yet fixed in Lidarr.
`mv ${SYNOPKG_PKGDEST}/app/config.xml ${LIDARR_CONFIG_DIR}/config.xml ` must be removed
See #5514 for the solution with Lidarr.

hgy59 commented:
@mreid-tt please can you test the Lidarr approach with sonarr?
unfortunately the `update_required` file is not deleted in Lidarr when updating with the current version of  #5514 😖

mreid-tt commented:
This PR will be closed and changes moved to a PR based on a unique branch in #5515
